lib/mfx/api.py

```diff
@@ -0,0 +1,141 @@
+    return self.ifprop("profile", inner)
+
+  @property
+  def rcmode(self):
+    return f"RateControlMethod={self.map_rcmode(self.props['rcmode'])}"
+
+  @property
+  def level(self):
+    def inner(level):
+      # TODO: need to define map_level
+      return f"CodecLevel={self.map_level(self.props['level'])}"
+    return self.ifprop("level", inner)
+
+  @property
+  def qp(self):
+    def inner(qp):
+      rcmode = self.props["rcmode"].upper()
+      mqp = qp
+      if self.codec in [Codec.MPEG2]:
+        mqp = mapRangeInt(qp, [0, 100], [1, 51])
+      elif self.codec in [Codec.AV1] and "ICQ" == rcmode:
+        mqp = mapRangeInt(qp, [0, 255], [1, 51])
+      elif self.codec in [Codec.HEVC] and "QVBR" == rcmode:
+        mqp = mapRangeInt(qp, [0, 255], [1, 51])
+      return f"QPI={mqp}:QPP={mqp}:QPB={mqp}"
+    return self.ifprop("qp", inner)
+
+  @property
+  def encparams(self):
+    params = [
+      self.profile, self.rcmode, self.qp, self.level, self.gop, self.bframes,
+      self.slices, self.minrate, self.maxrate, self.refs, self.ladepth,
+      self.extbrc, self.ldb, self.strict, self.pict, self.quality, self.lowpower,
+      self.tilecols, self.tilerows, self.maxframesize,
+    ]
+    return ':'.join(v for v in params if len(v) != 0)
+
+  @classmethod
+  @memoize
+  def map_profile(cls, codec, profile):
+    # TODO: we could move this method to the enum class as "lookup"
+    # then call MfxCodecProfile.lookup(codec, profile)
+    return {
+      Codec.AVC     : {
+        "high"      : MfxCodecProfile.MFX_PROFILE_AVC_HIGH,
+        "main"      : MfxCodecProfile.MFX_PROFILE_AVC_MAIN,
+        "baseline"  : MfxCodecProfile.MFX_PROFILE_AVC_BASELINE,
+        "unknown"   : MfxCodecProfile.MFX_PROFILE_UNKNOWN
+      },
+      Codec.HEVC   : {
+        "main"        : MfxCodecProfile.MFX_PROFILE_HEVC_MAIN,
+        "main444"     : MfxCodecProfile.MFX_PROFILE_HEVC_REXT,
+        "scc"         : MfxCodecProfile.MFX_PROFILE_HEVC_SCC,
+        "scc-444"     : MfxCodecProfile.MFX_PROFILE_HEVC_SCC,
+        "mainsp"      : MfxCodecProfile.MFX_PROFILE_HEVC_MAINSP,
+        "main10"      : MfxCodecProfile.MFX_PROFILE_HEVC_MAIN10,
+        "main10sp"    : MfxCodecProfile.MFX_PROFILE_HEVC_MAINSP,
+        "main444-10"  : MfxCodecProfile.MFX_PROFILE_HEVC_REXT,
+        "unknown"     : MfxCodecProfile.MFX_PROFILE_UNKNOWN
+      },
+      Codec.AV1 : {
+        "main"  : MfxCodecProfile.MFX_PROFILE_AV1_MAIN,
+      },
+      Codec.VP9 : {
+        "profile0"  : MfxCodecProfile.MFX_PROFILE_VP9_0,
+        "profile1"  : MfxCodecProfile.MFX_PROFILE_VP9_1,
+        "profile2"  : MfxCodecProfile.MFX_PROFILE_VP9_2,
+        "profile3"  : MfxCodecProfile.MFX_PROFILE_VP9_3
+      },
+    }[codec][profile]
+
+  @classmethod
+  @memoize
+  def map_rcmode(cls, rcmode):
+    # TODO: we could move this method to the enum class as "lookup"
+    # then call MfxRateControlMethod.lookup(rcmode)
```

Comment:
Please resolve this TODO.

Reply:
the next patch will move map_rate  to the enum class as lookup